types/context.go

```diff
@@ -275,6 +275,21 @@ func (c Context) CacheContext() (cc Context, writeCache func()) {
 	return cc, cms.Write
 }
 
+// CacheContextEmitEvents behaves the same as CacheContext except that it
+// automatically emits the events from the cached context when the caller
+// executes the writeCache function.
+func (c Context) CacheContextEmitEvents() (cc Context, writeCache func()) {
```

Comment:
Can we replace CacheContext instances inside the SDK to this new one, wherever relevant?